### PR TITLE
add services bridging to parameter_bridge

### PR DESCRIPTION
**Pull request by cyrilleberger**
making possible to define services using the parameter_bridge, it is done very similarly to topics:
It needs to parameter 'services_1_to_2' (for calling ROS2 services from ROS1) and 'services_2_to_1 (for calling ROS1 from ROS2), I tried to follow the naming convention of the dynamic_bridge. The parameters should look like:
[
{service: /name/of/service package: name_of_package_with_service, type: type_of_message},
...
]
For calling ROS1 add_two_ints from ROS2:
[
{service: /add_two_ints package: roscpp_tutorials, type: TwoInts}
]

### CODE REVIEW

File: src/parameter_bridge.cpp [hunk 1]
```diff
@@ -85,9 +95,93 @@ int main(int argc, char * argv[])
...
+  if (
+    ros1_node.getParam(services_1_to_2_parameter_name, services_1_to_2) &&
+    services_1_to_2.getType() == XmlRpc::XmlRpcValue::TypeArray)
+  {
+    for (size_t i = 0; i < static_cast<size_t>(services_1_to_2.size()); ++i) {
+      std::string service_name = static_cast<std::string>(services_1_to_2[i]["service"]);
+      std::string package_name = static_cast<std::string>(services_1_to_2[i]["package"]);
+      std::string type_name = static_cast<std::string>(services_1_to_2[i]["type"]);
+      printf(
+        "Trying to create bridge for ROS1 service '%s' "
```
Comment (dirk-thomas) [inline src/parameter_bridge.cpp]: The later called function `service_bridge_1_to_2` creates a ROS 2 client and a ROS 1 server. Shouldn't the message then say "bridge got ROS 2 service"?

Same inversion below.

File: src/parameter_bridge.cpp [hunk 2]
```diff
@@ -30,7 +30,6 @@
 
 #include "ros1_bridge/bridge.hpp"
 
-
```
Comment (dirk-thomas) [inline src/parameter_bridge.cpp]: Please avoid unrelated changes like this.

File: src/parameter_bridge.cpp [hunk 3]
```diff
@@ -67,7 +77,7 @@ int main(int argc, char * argv[])
         queue_size = 100;
       }
       printf(
-        "Trying to create bidirectional bridge for topic ''%s' "
+        "Trying to create bidirectional bridge for topic '%s' "
```
Comment (dirk-thomas) [inline src/parameter_bridge.cpp]: In the future please consider to create separate PRs for unrelated changes / fixes like this.

File: src/parameter_bridge.cpp [hunk 4]
```diff
@@ -85,9 +95,93 @@ int main(int argc, char * argv[])
...
+      auto factory = ros1_bridge::get_service_factory(
+        "ros2", package_name, type_name);
+      if (factory) {
+        try {
+          service_bridges_1_to_2.push_back(factory->service_bridge_1_to_2(ros1_node, ros2_node, service_name));
+          printf("Created 1 to 2 bridge for service %s\n", service_name.c_str());
+        } catch (std::runtime_error & e) {
+          fprintf(
+            stderr,
+            "failed to create bridge ROS1 service '%s' "
```
Comment (dirk-thomas) [inline src/parameter_bridge.cpp]: In the printed messages please use `ROS 1` / `ROS 2` (with a space between `ROS` and the number) as in the existing code.

**Comment by nuclearsandwich** [comment]
@cyrilleberger are you interested in resuming this pull request and #175?

**Comment by cyrilleberger** [comment]
Yes, not sure what is holding this one. #175 still has conflict, but I don't have anything to test #175 without having #176 and merging all the patch in one branch for testing is too messy, that is why I was hoping to wait for this patch to be merged before I do anything with #175.

**Comment by dirk-thomas** [comment]
> In the future please consider to create separate PRs for unrelated changes / fixes like this.

Removing the unrelated fix from the PR but not opening a separate PR is likely going to make this lost. I just created #207 for it - please do so on your own in the future.

File: src/parameter_bridge.cpp [hunk 5]
```diff
@@ -85,9 +96,93 @@ int main(int argc, char * argv[])
...
+  // ROS 2 Services in ROS 1
+  XmlRpc::XmlRpcValue services_2_to_1;
+  if (
+    ros1_node.getParam(services_2_to_1_parameter_name, services_2_to_1) &&
+    services_2_to_1.getType() == XmlRpc::XmlRpcValue::TypeArray)
+  {
+    for (size_t i = 0; i < static_cast<size_t>(services_2_to_1.size()); ++i) {
+      std::string service_name = static_cast<std::string>(services_2_to_1[i]["service"]);
+      std::string package_name = static_cast<std::string>(services_2_to_1[i]["package"]);
+      std::string type_name    = static_cast<std::string>(services_2_to_1[i]["type"]);
```
Comment (dirk-thomas) [inline src/parameter_bridge.cpp]: Please remove the alignment (as mentioned before in https://github.com/ros2/ros1_bridge/pull/176#discussion_r268235022).

**Comment by dirk-thomas** [comment]
@cyrilleberger Please rebase this against the latest state of the target branch in order to be able to run CI.

**Comment by dirk-thomas** [comment]
CI build: [Build Status](https://ci.ros2.org/job/ci_packaging_linux/225/)

Please address the reported linter warnings. Also make sure to run the tests of the package locally before submitting patches.

**Comment by cyrilleberger** [comment]
I have now fixed all the spaces, comma and so... I have three tests failing locally "xmllint", "test_topics_across_dynamic_bridge__rmw_fastrtps_cpp", "test_services_across_dynamic_bridge__rmw_fastrtps_cpp", but no clue why they fail and how they relate to my change.

**Comment by dirk-thomas** [comment]
Thanks for the patch / update. The CI build is looking good: [Build Status](https://ci.ros2.org/job/ci_packaging_linux/226/) (unrelated compiler warning also present on master)
